### PR TITLE
Add polynomial interpolation and ability to read RotNS data

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Add common polynomial interpolation code where the degree is a template parameter to eliminate memory allocations
- Minor cleanups related to clang-tidy
- Add ability to read RotNS data. This is created by the RotNS code described in the Cook, Shapiro, Teukolsky 92 and 94 papers. It allows for constructing equilibrium differentially and uniformly rotating NSs. In the future we will probably want to add density perturbations and magnetic fields, but those technically will require solving the constraint equations for the new metric, in which case having the elliptic solver just create the full initial data would be nicer.
- Add GRMHD executable for rotating NS. While this technically is just hydro, we don't have subcell for hydro only, and ultimately things are much more interesting with magnetic fields...

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/NumericalAlgorithms/Interpolation/PolynomialInterpolation.hpp [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+#include "Utilities/Gsl.hpp"
+
+namespace intrp {
+/*!
+ * \brief Interpolate `y_values` to `target_x` from tabulated `x_values` using a
+ * polynomial interpolant of degree `Degree`.
+ */
+template <size_t Degree>
+void polynomial_interpolation(gsl::not_null<double*> y,
+                              gsl::not_null<double*> error_in_y,
```
Comment (wthrowe) [inline src/NumericalAlgorithms/Interpolation/PolynomialInterpolation.hpp:23]: Document `error_in_y`.

File: src/NumericalAlgorithms/Interpolation/PolynomialInterpolation.cpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  size_t ns = index_of_closest_point;
+  *y = y_values[ns--];
+  for (size_t m = 1; m < order; m++) {
+    for (size_t i = 0; i < order - m; i++) {
+      const double delta_x_o = x_values[i] - target_x;
+      const double delta_x_p = x_values[i + m] - target_x;
+      const double c_minus_d =
+          gsl::at(algorithm_c, i + 1) - gsl::at(algorithm_d, i);
+      ASSERT(delta_x_o - delta_x_p != 0.0,
+             "Uniformly spaced grid points or repeated grid points. i:"
```
Comment (wthrowe) [inline src/NumericalAlgorithms/Interpolation/PolynomialInterpolation.cpp]: If uniformly spaced points don't work that should be documented.  (Repeated is kind of obvious.)

File: tests/Unit/NumericalAlgorithms/Interpolation/Test_PolynomialInterpolation.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    y += pow(coords, i);
+    expected_y += pow(target_x, i);
+  }
+  double error_in_y = 0.0;
+  double target_y = 0.0;
+  intrp::polynomial_interpolation<Degree>(
+      make_not_null(&target_y), make_not_null(&error_in_y), target_x,
+      gsl::make_span(y.data(), Degree + 1),
+      gsl::make_span(coords.data(), Degree + 1));
+  CHECK_ITERABLE_APPROX(target_y, expected_y);
```
Comment (wthrowe) [inline tests/Unit/NumericalAlgorithms/Interpolation/Test_PolynomialInterpolation.cpp:31]: Check `error_in_y`.

File: src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/RotatingStar.cpp [hunk 4]
```diff
@@ -0,0 +1,957 @@
...
+  std::ifstream cst_file(filename);
+  std::string header_line;
+  std::getline(cst_file, header_line);
+  // Read as integers from file to be compatible with SpEC. Might not be
+  // necessary.
+  int num_radial_points = -1;
+  int num_angular_points = -1;
+  cst_file >> num_radial_points >> num_angular_points;
+  num_radial_points_ = static_cast<size_t>(num_radial_points);
+  num_angular_points_ = static_cast<size_t>(num_angular_points) + 1;
```
Comment (wthrowe) [inline src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/RotatingStar.cpp:45]: I think this line deserves a comment.  Why is it one more than the number read from the file?

File: src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/RotatingStar.cpp [hunk 5]
```diff
@@ -0,0 +1,957 @@
...
+                        << "\nangular_index: " << angular_index);
+  }
+  if (UNLIKELY(radius_[num_angular_points_ * radius_index + angular_index] >
+               target_radius)) {
+    // This sanity check and adjustment is in SpEC, but Nils Deppe isn't sure
+    // under what circumstances it is necessary. Best guess is that because the
+    // radius_index is an estimate it could be that you compute an index that's
+    // slightly an over-estimate. We could do a binary search to find the exact
+    // index if the estimate fails.
+    // radius_index--;
```
Comment (wthrowe) [inline src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/RotatingStar.cpp]: Leaving this code commented is fine if we are unsure if it is needed, but comment the whole section (including the `if`).

File: src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/RotatingStar.hpp [hunk 6]
```diff
@@ -0,0 +1,574 @@
...
+ *
+ * \f{align}{
+ *   u_x
+ *   &=-\frac{e^{(\gamma-\rho)/2}v\sin(\phi)}{\sqrt{1-v^2}} \\
+ *   u_y
+ *   &=\frac{e^{(\gamma-\rho)/2}v\cos(\phi)}{\sqrt{1-v^2}} \\
+ *   u_z&=0.
+ * \f}
+ *
+ * \warning Near (within `1e-2`) $r=0$ the numerical errors from interpolation
```
Comment (wthrowe) [inline src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/RotatingStar.hpp]: `\f$r=0\f$`

File: src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/RotatingStar.cpp [hunk 7]
```diff
@@ -0,0 +1,957 @@
...
+      sin_theta(get_size(radius)),
+      delta_r(in_delta_r) {
+  detail::compute_angular_coordinates(
+      make_not_null(&radius), make_not_null(&cos_theta),
+      make_not_null(&sin_theta), make_not_null(&phi), coords);
+}
+
+template <typename DataType>
+auto RotatingStar::make_metric_data(const size_t num_points) const ->
+    typename IntermediateVariables<DataType>::MetricData {
```
Comment (wthrowe) [inline src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/RotatingStar.cpp]: Could this be a constructor for `MetricData`?

**Review by wthrowe — CHANGES_REQUESTED** [review]
I did not go through all the math in detail, but I expect any errors there would have been pretty obvious when the code was used.

**Comment by nilsdeppe** [comment]
rebase and pushed fixup commits. Thanks for the review @wthrowe!

**Comment by wthrowe** [comment]
Looks good.  Squash.

**Comment by nilsdeppe** [comment]
Squashed! Thanks!

**Review by wthrowe — DISMISSED** [review]

**Comment by mduez** [comment]
I am confused by this comment in the documentation:
"where the rotation is about the y-axis. We switch to rotation about the z-axis while switching to Cartesian components below."
RotNS does not, I believe, invoke Cartesian coordinates at all.  SpEC does take the x-axis as the rotation axis by default, but there is nothing imposing rotation axis = y in the RotNS output that needs to be rotated.  Can't you just say that your code takes the rotation axis to be the z-axis?

**Comment by mduez** [comment]
I believe your definition of v is the Eulerian velocity rather than the transport velocity.  If that's standard in SpECTRE, there's no need to note this.  In SpEC, v is the transport velocity, and we need to be careful to specify which velocity we mean--especially when it's a different one.

**Comment by nilsdeppe** [comment]
Rebased and squashed in rewording to "We use rotation about the \f$z\f$-axis." 

We use `v^i` to be the Eulerian velocity. I forget what exactly we use for the transport velocity (`v^i_t` maybe), but they're distinct.

File: src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/RotatingStar.hpp [hunk 8]
```diff
@@ -0,0 +1,576 @@
...
+ *
+ * We can transform from spherical to Cartesian coordinates using
+ *
+ * \f{align}{
+ *   \label{eq:Jacobian}
+ *   \frac{\partial (r,\theta,\phi)}{\partial (x,y,z)}=
+ *   \begin{pmatrix}
+ *     \cos(\phi) \sin(\theta) & \sin(\theta)\sin(\phi) & \cos(\theta) \\
+ *     \tfrac{\cos(\phi)\cos(\theta)}{r} & \tfrac{\sin(\phi)\cos(\theta)}{r}
+ *     & \tfrac{\cos(\phi)}{r} \\
```
Comment (kidder) [inline src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/RotatingStar.hpp]: should be `-sin(theta)/r`

File: src/NumericalAlgorithms/Interpolation/PolynomialInterpolation.cpp [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+                                       << x_values.size());
+  ASSERT(y_values.size() == order, "The x_values span must be of size "
+                                       << order << " but got "
+                                       << x_values.size());
+  using std::abs;
+  // C and D are constants used in the interpolation algorithm. They're an
+  // intermediate variable to make life easier.
+  std::array<double, order> algorithm_c{};
+  std::array<double, order> algorithm_d{};
+  const size_t num_points = x_values.size();
```
Comment (kidder) [inline src/NumericalAlgorithms/Interpolation/PolynomialInterpolation.cpp]: you assert above that this is equal to `order` so why introduce a new variable?

File: docs/References.bib [hunk 10]
```diff
@@ -448,6 +448,32 @@ @book{Cockburn1999
...
+                  Effects of General Relativity}",
+  journal =      {The Astrophysical Journal},
+  year =         1992,
+  month =        oct,
+  volume =       {398},
+  pages =        {203},
+  doi =          {10.1086/171849},
+}
+
+@ARTICLE{Cook1994,
```
Comment (kidder) [inline docs/References.bib:465]: these aren't generating links to the papers

File: src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/RotatingStar.cpp [hunk 11]
```diff
@@ -0,0 +1,957 @@
...
+  const size_t num_points = get_size(get<0>(coords));
+  *radius = get(magnitude(coords));
+  if constexpr (not std::is_fundamental_v<DataType>) {
+    cos_theta->destructive_resize(num_points);
+    sin_theta->destructive_resize(num_points);
+    phi->destructive_resize(num_points);
+  }
+  for (size_t i = 0; i < num_points; ++i) {
+    if (get_element(*radius, i) < 1.e-12) {
+      get_element(*cos_theta, i) = 0.5;
```
Comment (kidder) [inline src/PointwiseFunctions/AnalyticSolutions/RelativisticEuler/RotatingStar.cpp:336]: why is this theta chosen?

**Review by kidder — CHANGES_REQUESTED** [review]
you can squash

File: docs/References.bib (hunk 10 above)
Comment (nilsdeppe) [reply docs/References.bib:465]: I don't know of any online verison of these papers

**Comment by nilsdeppe** [comment]
rebased and squashed. Thanks for the reviews!

**Review by kidder — APPROVED** [review]

**Pull request merged**